daemon/api_sideload_n_try.go

```diff
@@ -218,39 +219,153 @@ func sideloadOrTrySnap(ctx context.Context, c *Command, body io.ReadCloser, boun
 
 // sideloadedInfo contains information from a bunch of sideloaded snaps
 type sideloadedInfo struct {
-	sideInfos                  []*snap.SideInfo
-	names, origPaths, tmpPaths []string
+	snaps      []sideloadSnapInfo
+	components []sideloadComponentInfo
 }
 
-func sideloadInfo(st *state.State, snapFiles []*uploadedSnap, flags sideloadFlags) (*sideloadedInfo, *apiError) {
+type sideloadSnapInfo struct {
+	sideInfo   *snap.SideInfo
+	components []sideloadComponentInfo
+	origPath   string
+	tmpPath    string
+}
+
+type sideloadComponentInfo struct {
+	sideInfo *snap.ComponentSideInfo
+	origPath string
+	tmpPath  string
+}
+
+func sideloadInfo(st *state.State, uploads []*uploadedContainer, flags sideloadFlags) (*sideloadedInfo, *apiError) {
 	deviceCtx, err := snapstate.DevicePastSeeding(st, nil)
 	if err != nil {
 		return nil, InternalError(err.Error())
 	}
 
-	names := make([]string, len(snapFiles))
-	origPaths := make([]string, len(snapFiles))
-	tmpPaths := make([]string, len(snapFiles))
-	sideInfos := make([]*snap.SideInfo, len(snapFiles))
+	var components []sideloadComponentInfo
+	var snaps []sideloadSnapInfo
+	for _, upload := range uploads {
+		si, snapErr := readSideInfo(st, upload.tmpPath, upload.filename, flags, deviceCtx.Model())
+		if snapErr != nil {
+			if !flags.dangerousOK {
+				// TODO:COMPS: read assertions for components
```

Comment:
ah there is still TODO here about that

Reply:
Yeah, I've got a branch ready for once this is merged to address this TODO.